pkg/controllers/taint/clustertaintpolicy_controller.go

```diff
@@ -0,0 +1,246 @@
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (c *ClusterTaintPolicyController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling Cluster %s.", req.Name)
+
+	clusterObj := &clusterv1alpha1.Cluster{}
+	if err := c.Client.Get(ctx, req.NamespacedName, clusterObj); err != nil {
+		klog.Errorf("Failed to get Cluster(%s), error: %v", req.Name, err)
+		return controllerruntime.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !clusterObj.DeletionTimestamp.IsZero() {
+		klog.V(4).Infof("Cluster(%s) is deleting.", req.Name)
+		return controllerruntime.Result{}, nil
+	}
+
+	clusterTaintPolicyList := &policyv1alpha1.ClusterTaintPolicyList{}
+	listOption := &client.ListOptions{UnsafeDisableDeepCopy: ptr.To(true)}
+	if err := c.Client.List(ctx, clusterTaintPolicyList, listOption); err != nil {
+		klog.Errorf("Failed to list ClusterTaintPolicy, error: %v", err)
+		return controllerruntime.Result{}, err
+	}
+
+	clusterCopyObj := clusterObj.DeepCopy()
+	for _, policy := range clusterTaintPolicyList.Items {
+		if policy.Spec.TargetClusters != nil && !util.ClusterMatches(clusterCopyObj, *policy.Spec.TargetClusters) {
+			continue
+		}
+
+		clusterCopyObj.Spec.Taints = taintClusterWithPolicy(clusterCopyObj, policy)
+	}
+
+	if !reflect.DeepEqual(clusterObj.Spec.Taints, clusterCopyObj.Spec.Taints) {
+		objPatch := client.MergeFrom(clusterObj)
+		err := c.Client.Patch(ctx, clusterCopyObj, objPatch)
+		if err != nil {
+			klog.Errorf("Failed to patch Cluster(%s), error: %v", req.Name, err)
+			c.EventRecorder.Event(clusterCopyObj, corev1.EventTypeWarning, events.EventReasonTaintClusterFailed, err.Error())
+			return controllerruntime.Result{}, err
+		}
+
+		msg := fmt.Sprintf("Taint Cluster(%s) with taints(%v) succeed", req.Name, utilhelper.GenerateTaintsMessage(clusterCopyObj.Spec.Taints))
+		klog.V(4).Info(msg)
+		c.EventRecorder.Event(clusterCopyObj, corev1.EventTypeNormal, events.EventReasonTaintClusterSucceed, msg)
+		return controllerruntime.Result{}, nil
+	}
+	klog.V(4).Infof("Cluster(%s) taints are up to date.", req.Name)
+	return controllerruntime.Result{}, nil
+}
+
+func taintClusterWithPolicy(cluster *clusterv1alpha1.Cluster, policy policyv1alpha1.ClusterTaintPolicy) []corev1.Taint {
+	if conditionMatches(cluster, policy.Spec.MatchConditions) {
+		return addTaintsOnCluster(cluster, policy.Spec.Taints, metav1.Now())
+	}
+	return removeTaintsOnCluster(cluster, policy.Spec.Taints)
+}
+
+// conditionMatches returns true if the cluster matches the match conditions.
+// The match conditions are ANDed.
+// If the match conditions are empty, it will return false.
+func conditionMatches(cluster *clusterv1alpha1.Cluster, matchConditions []policyv1alpha1.MatchCondition) bool {
```

Comment:
```suggestion
func conditionMatches(Conditions []metav1.Condition, matchConditions []policyv1alpha1.MatchCondition) bool {
```
update the comments accordingly if it is accepted.